src/lib/core/ordered_latch.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include <assert.h>
+#include "fiber.h"
+#include "small/rlist.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * Latch of cooperative multitasking environment, which preserves strict order
+ * of fibers waiting for the latch.
+ */
+struct ordered_latch {
+	/** The fiber that locked the latch, or NULL if unlocked. */
+	struct fiber *owner;
+	/** The queue of fibers waiting for the latch. */
+	struct rlist queue;
+};
+
+/**
+ * Initialize the given latch.
+ */
+static inline void
+ordered_latch_create(struct ordered_latch *l)
+{
+	l->owner = NULL;
+	rlist_create(&l->queue);
+}
+
+/**
+ * Destroy the given latch.
+ */
+static inline void
+ordered_latch_destroy(struct ordered_latch *l)
+{
+	assert(l->owner == NULL);
+	assert(rlist_empty(&l->queue));
+	(void)l;
+}
+
+/**
+ * Return true if the latch is locked.
+ */
+static inline bool
+ordered_latch_is_locked(const struct ordered_latch *l)
+{
+	return l->owner != NULL;
+}
+
+/**
+ * Lock the latch.
+ * If it is already locked by another fiber, wait for it.
+ * Locks are acquired in the strict order as they were requested.
+ */
+static inline void
+ordered_latch_lock(struct ordered_latch *l)
+{
+	assert(l->owner != fiber());
+	if (l->owner == NULL && rlist_empty(&l->queue)) {
+		l->owner = fiber();
+		return;
+	}
+
+	struct ordered_latch waiter;
+	waiter.owner = fiber();
+	rlist_add_tail_entry(&l->queue, &waiter, queue);
+	do {
+		fiber_yield();
+	} while (l->owner != NULL);
```

Comment:
You made it so that waking a fiber up doesn't put it in front of the queue, but waking up some fiber from the end of the queue will still make it steal the latch. I think there should be some check for whether this waiter is the first in queue or not.
The test fails after the following change:
```diff
diff --git a/test/unit/ordered_latch.c b/test/unit/ordered_latch.c
index 79dd6eea0..a455a99a7 100644
--- a/test/unit/ordered_latch.c
+++ b/test/unit/ordered_latch.c
@@ -36,13 +36,13 @@ main_f(va_list ap)
                fiber_set_joinable(fibers[i], true);
                fiber_start(fibers[i], &i, &check, &latch);
        }
-
-       ordered_latch_unlock(&latch);
        /*
         * Try to break the order of waiters on the latch.
         */
        fiber_wakeup(fibers[1]);
 
+       ordered_latch_unlock(&latch);
+
        for (size_t i = 0; i < num_fibers; i++)
                fiber_join(fibers[i]);
        ordered_latch_destroy(&latch);
```
```
[001] unit/ordered_latch.test                                         [ fail ]
[001] 
[001] Test failed! Result content mismatch:
[001] --- unit/ordered_latch.result	Fri Jul  8 09:49:30 2022
[001] +++ /tmp/t/rejects/unit/ordered_latch.reject	Fri Jul  8 10:02:15 2022
[001] @@ -1,7 +1,12 @@
[001] +#   Failed test 'check order'
[001] +#   in /Users/s.petrenko/Source/sptnt/test/unit/ordered_latch.c at line 17
[001] +#   Failed test 'check order'
[001] +#   in /Users/s.petrenko/Source/sptnt/test/unit/ordered_latch.c at line 17
[001] +# Looks like you failed 2 test of 3 run.
[001]  1..3
[001] +fiber 1 acquired the latch
[001] +not ok 1 - check order
[001]  fiber 0 acquired the latch
[001] -ok 1 - check order
[001] -fiber 1 acquired the latch
[001] -ok 2 - check order
[001] +not ok 2 - check order
[001]  fiber 2 acquired the latch
[001]  ok 3 - check order
[Main process] Got failed test; gently terminate all workers...
[001] Worker "001_unit" got failed test; stopping the server...```

Reply:
Thanks for the catch.